feat: openapi typescript client (#2885)
Redocly/redocly-cli: docs/@v2/commands/generate-client.md

Reviewer: add please to `docs/@v2/v2.sidebars.yaml`

Author: done

Reviewer: could we split this into two pages: CLI command usage (flags, config, output modes) and client usage (auth, retries, error handling, middleware, the extra generators) ?

Author: make sense, in progress

Author: done
commands/generate-client.md
guides/use-generated-client.md
configuration/reference/client.md

---
slug:
  - /docs/cli/commands/generate-client
---

# `generate-client`

{% admonition type="warning" name="Experimental" %}
`generate-client` is **experimental**.

> Its CLI flags, generated output, configuration schema, and the custom-generator plugin API may change in any minor release until the feature is declared stable.

Reviewer 2: ```suggestion
Its interface, generated output, and configuration schema may change in any release until the feature is declared stable.
```

Author: corrected

> Pin your `@redocly/cli` version if you depend on the generated output, and plan to regenerate when you upgrade.

Reviewer 2: It should be obvious from the above sentence.

Author: corrected

We'd love your feedback while we stabilize it.
{% /admonition %}

> Generate a typed TypeScript client from an OpenAPI description.

Reviewer 2: Could you follow the same structure as in other commands' docs? And it would be nice if we can shorten the description or restrutcure / split it into several pages as it's easy to get lost here.

Author: yeah, I made some changes in this file, but I think I will need to do final review again

> Accepts **OpenAPI 3.x**, plus **Swagger 2.0** (normalized to the 3.x shape before generation).

Reviewer 2: I'm not even sure it worth supporting Swagger at this point. Most likely all its audience already use something. I'd target new users (hence the recent standards) mostly.

Author: Well, it supports now, why not mention it?

Reviewer 2: Because of the maintenance burden. But maybe it's not that important at this point -- let it be.

`<input>` is a file path, a URL, or an `apis:` alias from `redocly.yaml`.

The generated client has **zero runtime dependencies** — it uses only web-standard APIs (`fetch`, `AbortController`, `URLSearchParams`, …), so it runs in browsers, Node, Bun, Deno, and edge runtimes.
Code is produced through the TypeScript compiler AST (not string templates), so output is correct by construction.
By default it emits a single file containing inline types and one async function per operation.

## Usage

```sh

> npx @redocly/cli@latest generate-client <input> --output <file.ts>

Reviewer 2: It's better to use one notation (either `npx @redocly/cli ...` or  simply `redocly ...`) but not both in the same example as it looks confusing. (I prefer the second one for consistency with other docs.)

Author: changed to redocly ...

> # <input> is an OpenAPI description file path or an `apis:` alias from redocly.yaml

Reviewer 2: This is a duplication of https://github.com/Redocly/redocly-cli/pull/2885/changes#diff-b7cda341b3c97c80168c0afe7d149a7309ea9da7b84d11cc2963e2a7898cae63R18. And I'd use <api>/<apis> like in other commands.

Author: corrected a bit

redocly generate-client openapi.yaml --output src/client.ts
redocly generate-client cafe -o src/client.ts
```

## Options

> {% table %}

Reviewer 2: Could you use regular table here? We use Markdoc and Markdown tables here and there, but I'd like to make that uniform. And  I don't see much benefits from the Marcdoc one over regular tables, so we'll probably convert them anyway to be uniform.

Author: changed to regular table

---

- `--enum-style`
- `string`
- How named string enums are emitted: `const-object` (default) emits a runtime `as const` object alongside the union type; `union` emits only the union type.

---

- `--error-mode`
- `string`
- Error-handling shape: `throw` (default) throws `ApiError` on a non-2xx response.
  `result` returns a discriminated `{ data, error, response }` object instead, with `error` typed from the spec's 4xx/5xx response bodies.

---

- `--base-url`
- `string`
- Override the `BASE` URL inlined into the generated runtime.
  Defaults to `servers[0].url` from the description. Must be a valid URL.

---

- `--setup`
- `string`
- Path to a publisher setup module (`export default defineClientSetup({ config, middleware })`) baked into the generated client, so a published SDK ships its request/response defaults built in. Works across all output modes and both facades. See [Bake defaults into a published SDK](#bake-defaults-into-a-published-sdk---setup).

---

- `--config`
- `string`
- Path to the `redocly.yaml` configuration file (where the `x-client-generator` block lives).
  Defaults to the `redocly.yaml` discovered in the working directory.

{% /table %}

> ## Configuration

Reviewer: maybe we can move it to `docs/@v2/configuration` ? 
wdyt

Author: make sense, in progress

Author: done
commands/generate-client.md
guides/use-generated-client.md
configuration/reference/client.md

Instead of passing flags every time, you can put the settings in your `redocly.yaml` under an `x-client-generator` block.
`generate-client` reads it automatically — from a `redocly.yaml` in the working directory, or one pointed to by the standard `--config` flag:

```yaml
# redocly.yaml
x-client-generator:
  input: ./openapi.yaml
  output: ./src/api/client.ts
  generators:
    - sdk
  facade: service-class
```

Then run:

```sh
redocly generate-client
```

Relative `input`/`output` are resolved against the `redocly.yaml` directory, so the command works the same from any working directory.
Point at a `redocly.yaml` elsewhere with the standard `--config` flag:

```sh
redocly generate-client --config ./config/redocly.yaml
```

> **Precedence** (lowest to highest): the `redocly.yaml` `x-client-generator` block → individual CLI flags.

Reviewer 2: It's the same precedence order as for other commands (CLI flags -> config), and this is  a common approach. So there's no need to state that separately, I believe.

Author: removed

```sh
redocly generate-client openapi.yaml --output src/client.ts \
  --generators sdk,mock --mock-data faker --mock-seed 42
```

{% admonition type="info" name="Compatibility" %}
`mock` requires `--generators sdk` (the factories reference its types).
Install MSW in your app as a dev dependency (`msw` `^2`).
For `--mock-data faker`, also install `@faker-js/faker`.
The generated client itself stays dependency-free — only the `*.mocks.ts` module imports them.
{% /admonition %}

## Custom generators

The built-in generators (`sdk`, `zod`, …) cover the common targets.
For anything else derived from the same description — validators in another library, a permissions map for your UI, mocks in your test runner's format, an SDK in your company's house style — write a **custom generator**.
It reads the same spec-agnostic model the built-ins consume and runs in the same command, so its output never drifts from the spec and you never parse OpenAPI yourself.

A generator is `{ name, run }` (plus optional compatibility metadata).
`run` receives the model and returns files; author it with `defineGenerator` for type inference:

```ts
// route-map-generator.ts

> import { defineGenerator } from '@redocly/client-generator/plugin';

Reviewer 2: Can we avoid introducing a duplicated plugin concept? If needed, let's adjust our existing plugin format to accept the generator's part.

Author: yes we can, changed logic

---
slug:
  - /docs/cli/commands/generate-client
---

# `generate-client`

{% admonition type="warning" name="Experimental" %}
`generate-client` is **experimental**: its flags, generated output, configuration schema, and custom-generator API may change in any minor release until it's stable.
We'd love your feedback while we stabilize it.
{% /admonition %}

Generate a typed TypeScript client from an OpenAPI 3.x description.
Swagger 2.0 descriptions are also accepted — normalized to the 3.x shape before generation.

The generated client has **zero runtime dependencies** by default — it uses only web-standard APIs (`fetch`, `AbortController`, `URLSearchParams`, …), so it runs in browsers, Node, Bun, Deno, and edge runtimes.
By default it emits a single self-contained file with inline types and one async function per operation; opt in to a versioned, npm-updatable runtime instead with [`--runtime package`](#runtime-distribution).

This page covers running the command; for the generated client's runtime API (auth, error handling, middleware, retries, and the add-on generators) see [Use the generated client](../guides/use-generated-client.md).

## Usage

```sh

> redocly generate-client                          # every api with a `client` block (see Configuration)

Reviewer 2: >  every api with a `client` block

The above make this command stand out from other commands (like lint), that work out every `apis` section, merging the local config with the the root one. I'd lean towards producing an output to every `api` that has `clientOutput` defined instead.

redocly generate-client cafe                     # a single `apis:` alias from redocly.yaml

> redocly generate-client openapi.yaml -o src/client.ts   # a file path (ignores `apis:`)

Reviewer 2: ```suggestion
redocly generate-client openapi.yaml -o dist/client.ts   # a file path (ignores `apis:`)
```

src isn't a perfect folder name for a generated code.

| `--mock-data`       | `string` | `mock` generator data: `baked` (default, deterministic literals) or `faker` (`@faker-js/faker` calls).                                                                                                                            |
| `--mock-seed`       | `number` | Seed for `faker`-mode mocks, for reproducible data. Ignored in `baked` mode.                                                                                                                                                      |
| `--server-url`      | `string` | Override the server URL inlined into the runtime. Defaults to `servers[0].url`. Accepts absolute (`https://api.example.com`) or relative (`/v1`).                                                                                 |
| `--setup`           | `string` | Path to a publisher setup module baked into the client. See [Publisher defaults](../guides/use-generated-client.md#publisher-defaults).                                                                                           |

> | `--config`          | `string` | Path to the `redocly.yaml` (the `client` config and `apis:` live there). Defaults to the one in the working directory.                                                                                                            |

Reviewer 2: ```suggestion
| `--config`          | `string` | Path to the config file. Defaults to `redocly.yaml` in the working directory.                                                                                                            |
```

## Configuration

> Instead of passing flags every time, keep the settings in `redocly.yaml` under a top-level `client` block and per-API `apis.<name>.client` / `clientOutput`. Settings resolve **top-level `client` → per-API `client` → CLI flags**. See [`client` configuration](../configuration/reference/client.md) for the full reference.

Reviewer 2: > Settings resolve **top-level `client` → per-API `client` → CLI flags**.

this doesn't seem right. CLI args take precedence over config, and the config is resolved a bit differently (apis-level options override the root ones, as described [here](https://redocly.com/docs/cli/configuration/apis), or override the root ones completely -- depends on what we decide).

Author: corrected

| `--runtime`         | `string`   | Where the client's engine lives. See [Choose a runtime](#choose-a-runtime). <br/> **Possible values:** `inline`, `package`. Default value is `inline`.                                                                                                           |
| `--import-ext`      | `string`   | Extension in generated relative imports. See [Run with Node directly](../guides/use-generated-client.md#run-with-node-directly). <br/> **Possible values:** `js` (the tsc/bundler convention), `ts` (for Node's built-in type stripping). Default value is `js`. |
| `--generator`       | `[string]` | Generator to run — a built-in name or a custom generator's path or package; repeat the flag to run several. Default value is `sdk`. See [Generators](../guides/use-generated-client.md#generators).                                                              |
| `--args-style`      | `string`   | How operation inputs are passed. See [Argument style](../guides/use-generated-client.md#argument-style). <br/> **Possible values:** `flat`, `grouped`. Default value is `flat`.                                                                                  |

> | `--enum-style`      | `string`   | How named string enums are emitted. <br/> **Possible values:** `const-object` (an `as const` object + union), `union` (union only). Default value is `const-object`.                                                                                             |

Reviewer 3: do we need this knob? I would just use const object

Author: good catch, removed

| `--error-mode`      | `string`   | How operations report HTTP errors. See [Error handling](../guides/use-generated-client.md#error-handling). <br/> **Possible values:** `throw`, `result`. Default value is `throw`.                                                                               |
| `--date-type`       | `string`   | Type of `date`/`date-time` fields; pair `Date` with the `transformers` generator. <br/> **Possible values:** `string`, `Date`. Default value is `string`.                                                                                                        |

> | `--query-framework` | `string`   | Framework adapter for the `tanstack-query` generator. <br/> **Possible values:** `react`, `vue`, `svelte`, `solid`. Default value is `react`.                                                                                                                    |

Reviewer 3: should it be called then `--tanstack-query-framework` or alternatively can we have separate generators for `tanstack-query`: `tanstack-query-react`, `tanstack-query-vue` ...

Author: I like second option, will do it

Author: expect(importOf('tanstack-query')).toBe('@tanstack/react-query');
  expect(importOf('tanstack-query-vue')).toBe('@tanstack/vue-query');
  expect(importOf('tanstack-query-svelte')).toBe('@tanstack/svelte-query');
  expect(importOf('tanstack-query-solid')).toBe('@tanstack/solid-query');

Author: corrected

> | `--mock-data`       | `string`   | Data mode for the `mock` generator. <br/> **Possible values:** `baked` (deterministic literals), `faker` (`@faker-js/faker` calls). Default value is `baked`.                                                                                                    |

Reviewer 3: ```suggestion
| `--mock-data`       | `string`   | Data mode for the `mock` generator. <br/> **Possible values:** `static` (deterministic literals), `faker` (`@faker-js/faker` calls). Default value is `baked`.                                                                                                    |
```

Author: renamed

| `--mock-seed`       | `number`   | Seed for `faker`-mode mocks, for reproducible data. Ignored in `baked` mode.                                                                                                                                                                                     |

> | `--server-url`      | `string`   | Override the server URL baked into the client. Accepts an absolute (`https://api.example.com`) or relative (`/v1`) URL. Defaults to `servers[0].url`.                                                                                                            |

Reviewer 3: We should also support custom URL in the client constructor of the generated SDK (we may have a knob to allow/disallow it)

Author: good news we support this, just added  "The app can also repoint the client at runtime — createClient({ serverUrl }) or configure({ serverUrl }), see Authentication in the usage guide" to docs

---
slug:
  - /docs/cli/commands/generate-client
---

# `generate-client`

{% admonition type="warning" name="Experimental" %}
`generate-client` is **experimental**.
Its CLI flags, generated output, configuration schema, and the custom-generator plugin API may change in any minor release until the feature is declared stable.
Pin your `@redocly/cli` version if you depend on the generated output, and plan to regenerate when you upgrade.
We'd love your feedback while we stabilize it.
{% /admonition %}

Generate a typed TypeScript client from an OpenAPI description.

> Accepts **OpenAPI 3.0, 3.1, and 3.2.0**, plus **Swagger 2.0** (normalized to the 3.x shape before generation).

Reviewer 2: ```suggestion
Accepts **OpenAPI 3.x**, plus **Swagger 2.0** (normalized to the 3.x shape before generation).
```
